pkg/kubelet/nodestatus/setters.go

```diff
@@ -192,6 +174,55 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 	}
 }
 
+// shouldUseCloudProviderReportedHostname decides whether the given hostname or
+// the one reported by the cloud provider should be used
+func shouldUseCloudProviderReportedHostname(hostname string, // typically Kubelet.hostname, reported by the OS
```

Comment:
In the case where `cloudNodeAddresses` doesn't have a Hostname-type address, and also doesn't contain an address value matching `hostname`, and `hostnameOverridden` is false, this function will return `true`. That's really confusing, since the cloud provider didn't provide a hostname.

I think reversing all the checks and naming this function this way makes it more confusing. What we really want to check is whether we should use the given hostname or not. Naming this `shouldEnsureHostname(...)` and restoring the original polarity of the checks would be simpler to follow (it would also save negating the result in the only place this is called from)

Reply:
Actually, if this is only called in one place, I don't think breaking it out is that helpful... it loses the information about whether the addition was because of an override and should be warned about

Reply:
Yes, with my latest refactor it's called only once. I've put all this code back inline, now it should be a lot less confusing and the changeset smaller.

Universal feedback has been that this code is confusing, appreciate the help to make it readable!